File: java/core/src/main/java/com/github/pukkaone/mullet/html/PageBuilder.java
/*
Copyright (c) 2011, Chin Huang
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

  * Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package com.github.pukkaone.mullet.html;

import com.github.pukkaone.mullet.EscapeXml;
import com.github.pukkaone.mullet.html.parser.SimpleParser;
import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.ext.DefaultHandler2;

/**
 * Handles SAX events to extract data from rendered page.
 */
class PageBuilder extends DefaultHandler2 {
    private static final String HEAD = "head";
    private static final String TITLE = "title";
    private static final String BODY = "body";
    private static final char[] START_CDATA = "<![CDATA[".toCharArray();
    private static final char[] END_CDATA = "]]>".toCharArray();
    
    // number of open head elements
    private int headCount;
    
    // Count of nested open elements where this handler is extracting their
    // inner HTML.
    private int innerDepth;
    private StringBuilder innerHtml;
    private Page page = new Page();

    private boolean isExtractingInnerHtml() {
        return innerDepth > 0;
    }
    
    private void startExtractingInnerHtml() {
        innerDepth = 1;
        innerHtml = new StringBuilder();
    }
    
    private void append(char[] data, int start, int length) {
        innerHtml.append(data, start, length);
    }
    
    private PageBuilder append(char data) {
        innerHtml.append(data);
        return this;
    }
    
    private PageBuilder append(String data) {
        innerHtml.append(data);
        return this;
    }
    
    private String getInnerHtml() {
        return innerHtml.toString();
    }
    
    private void renderStartTag(String qualifiedName, Attributes attributes) {
        append('<').append(qualifiedName);

        for (int i = 0; i < attributes.getLength(); ++i) {
            append(' ').append(attributes.getQName(i)).append('=')
                    .append('"')
                    .append(EscapeXml.escape(attributes.getValue(i)))
                    .append('"');
        }
        
        append('>');
    }
    
    @Override
    public void startElement(
            String namespaceUri,
            String localName,
            String qualifiedName,
            Attributes attributes)
        throws SAXException
    {
        if (HEAD.equals(qualifiedName)) {
            ++headCount;
        }
        
        if (isExtractingInnerHtml()) {
            ++innerDepth;
            renderStartTag(qualifiedName, attributes);
            return;
        }
        
        if (headCount > 0 && TITLE.equals(qualifiedName)) {
            startExtractingInnerHtml();
            return;
        }
        
        if (BODY.equals(qualifiedName)) {
            startExtractingInnerHtml();
        }
    }
    
    private void renderEndTag(String qualifiedName) {
        append("</").append(qualifiedName).append('>');
    }
    
    @Override
    public void endElement(
            String namespaceUri, String localName, String qualifiedName)
        throws SAXException
    {
        if (HEAD.equals(qualifiedName)) {
            --headCount;
        }
        
        if (isExtractingInnerHtml()) {
            --innerDepth;
            if (isExtractingInnerHtml()
             && !SimpleParser.IMPLICIT_END_TAG_NS_URI.equals(namespaceUri)) {
                renderEndTag(qualifiedName);
            }
        }

        if (headCount > 0 && TITLE.equals(qualifiedName)) {
            page.title = getInnerHtml();
            return;
        }
        
        if (BODY.equals(qualifiedName)) {
            page.body = getInnerHtml();
        }
    }
    
    @Override
    public void characters(char[] ch, int start, int length)
        throws SAXException
    {
        if (isExtractingInnerHtml()) {
            append(ch, start, length);
        }
    }
    
    @Override
    public void startCDATA() throws SAXException {
        characters(START_CDATA, 0, START_CDATA.length);
    }
    
    @Override
    public void endCDATA() throws SAXException {
        characters(END_CDATA, 0, END_CDATA.length);
    }
    
    @Override
    public void ignorableWhitespace(char[] ch, int start, int length)
        throws SAXException
    {
        characters(ch, start, length);
    }
    
    @Override
    public void comment(char[] ch, int start, int length)
        throws SAXException
    {
        if (isExtractingInnerHtml()) {
            append("<!--");
            characters(ch, start, length);
            append("-->");
        }
    }
    
    @Override
    public void processingInstruction(String target, String data)
        throws SAXException
    {
        if (isExtractingInnerHtml()) {
            append("<?");
            append(target);
            append(" ");
            append(data);
            append("?>");
        }
    }
    
    Page getPage() {
        return page;
    }
}

File: ruby/core/lib/mullet/html/page_builder.rb
require 'mullet/html/attributes'
require 'mullet/html/command'
require 'mullet/html/element'
require 'mullet/html/element_renderer'
require 'mullet/html/for_element_renderer'
require 'mullet/html/if_element_renderer'
require 'mullet/html/parser/default_handler'
require 'mullet/html/parser/simple_parser'
require 'mullet/html/remove_mode'
require 'mullet/html/static_text_renderer'
require 'mullet/html/template'
require 'mullet/html/unless_element_renderer'
require 'mullet/template_error'
require 'set'

module Mullet; module HTML

  # Handles SAX events to extract content from an HTML page.
  class PageBuilder < Parser::DefaultHandler

    HEAD = 'head'
    TITLE = 'title'
    BODY = 'body'
    START_CDATA = '<![CDATA['
    END_CDATA = ']]>'

    # TODO: Write an alternative implementation where we don't have to know
    # which HTML elements have an empty content model.
    EMPTY_CONTENT_ELEMENTS = Set[
        'br',
        'hr',
        'img',
        'input']

    attr_reader :page

    def start_document()
      # number of open head elements
      @head_count = 0

      # Count of nested open elements where this handler is extracting their
      # inner HTML.
      @inner_depth = 0
    
      @page = Hash.new()
    end

    def start_element(name, attributes)
      if name == HEAD
        @head_count += 1
      end

      if extracting_inner_html?()
        @inner_depth += 1
        render_start_tag(name, attributes)
        return
      end

      if @head_count > 0 && name == TITLE
        start_extracting_inner_html()
        return
      end

      if name == BODY
        start_extracting_inner_html()
      end
    end

    def end_element(name)
      if name == HEAD
        @head_count -= 1
      end

      if extracting_inner_html?()
        @inner_depth -= 1
        if extracting_inner_html?() && !EMPTY_CONTENT_ELEMENTS.include?(name)
          render_end_tag(name)
        end
 
      end

      if @head_count > 0 && name == TITLE
        @page.store(:title, @inner_html)
        return
      end

      if name == BODY
        @page.store(:body, @inner_html)
      end
    end

    def characters(data)
      if extracting_inner_html?()
        append(data)
      end
    end

    def cdata_block(data)
      if extracting_inner_html?()
        append(data)
      end
    end

    def comment(data)
      if extracting_inner_html?()
        append("<!--#{data}-->")
      end
    end

    def processing_instruction(data)
      if extracting_inner_html?()
        append("<?#{data}?>")
      end
    end

    private

    def extracting_inner_html?()
      return @inner_depth > 0
    end

    def start_extracting_inner_html()
      @inner_depth = 1
      @inner_html = ''
    end

    def append(data)
      @inner_html << data
    end

    def escape_quote(value)
      return value.include?('"') ? value.gsub(/"/, '&#34;') : value
    end

    def render_start_tag(tag_name, attributes)
      append("<#{tag_name}")

      attributes.each do |attribute|
        qualified_name =
            [attribute.prefix, attribute.localname].compact().join(':')
        append(%( #{qualified_name}="#{escape_quote(attribute.value)}"))
      end

      append('>')
    end

    def render_end_tag(name)
      append("</#{name}>")
    end
  end

end; end
